File: android/src/main/java/net/infoxication/reactstarprnt/CallWithPermission.java
package net.infoxication.reactstarprnt;

import android.content.pm.PackageManager;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

import java.util.concurrent.Callable;

public class CallWithPermission {
  private AppCompatActivity activity;
  private String permission;
  private ActivityResultLauncher<String> requestPermissionActivity;
  private Boolean requestPermissionActivityResult;


  CallWithPermission(AppCompatActivity activity, String permission) {
    this.activity = activity;
    this.permission = permission;
  }

  public <O> O callMayThrow(Callable<O> grantedCallback, Runnable deniedCallback) throws Exception {
    if (ContextCompat.checkSelfPermission(this.activity, this.permission) == PackageManager.PERMISSION_GRANTED) {
      return grantedCallback.call();
    } else {
      this.requestPermissionActivity = this.activity.getActivityResultRegistry().register(
        "CallWithPermission_" + this.permission,
        new ActivityResultContracts.RequestPermission(),
        (Boolean result) -> {
          this.requestPermissionActivityResult = result;
        }
      );
      this.requestPermissionActivity.launch(this.permission);
      while (this.requestPermissionActivityResult == null) {
        try {
          Thread.sleep(500);
        } catch (InterruptedException e) {
          this.requestPermissionActivity.unregister();
        }
      }
      if (this.requestPermissionActivityResult) {
        return grantedCallback.call();
      } else {
        deniedCallback.run();
      }
    }
    return null;
  }
}

File: android/src/main/java/fr/wavyapp/reactNativeStarPrinter/CallWithPermission.kt
package fr.wavyapp.reactNativeStarPrinter

import android.content.pm.PackageManager
import androidx.activity.result.ActivityResultLauncher
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import java.util.concurrent.Callable

class CallWithPermission internal constructor(
  private val activity: AppCompatActivity,
  private val permission: String
) {
  private var requestPermissionActivity: ActivityResultLauncher<String>? = null
  private var requestPermissionActivityResult: Boolean? = null
  @Throws(Exception::class)
  fun <O> callMayThrow(grantedCallback: Callable<O>, deniedCallback: Runnable): O? {
    if (ContextCompat.checkSelfPermission(
        activity,
        permission
      ) == PackageManager.PERMISSION_GRANTED
    ) {
      return grantedCallback.call()
    } else {
      requestPermissionActivity = activity.activityResultRegistry.register(
        "CallWithPermission_$permission",
        ActivityResultContracts.RequestPermission()
      ) { result: Boolean? ->
        requestPermissionActivityResult = result
      }
      requestPermissionActivity!!.launch(permission)
      while (requestPermissionActivityResult == null) {
        try {
          Thread.sleep(500)
        } catch (e: InterruptedException) {
          requestPermissionActivity!!.unregister()
        }
      }
      if (requestPermissionActivityResult!!) {
        return grantedCallback.call()
      } else {
        deniedCallback.run()
      }
    }
    return null
  }
}
